isisd: add support of delay, TE metric and prefix-metric flag to flex-algorithm

louis-6wind opened the pull request.

gpziemba committed:
tests: add isis_sr_flex_algo_metrics topotests

Add isisd/zebra config files to set up flex-algo paths based
on te metric/delay.

Low te-metric path: 0,1,4,8,7,9
Low delay path: 0,5,6,2,3,9

Signed-off-by: G. Paul Ziemba <paulz@labn.net>
Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
doc: update flex-algo metrics in isisd user manual

Update the isisd user manual to document the support of flex-algo delay
and TE metric.

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: add metric-type TE and delay support to flex-algo

Add metric-type TE and delay support to flex-algo.

Signed-off-by: G. Paul Ziemba <paulz@labn.net>
Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: add support of flex-algo prefix-metric sub-TLV

Add the flex-algo prefix-metric sub-TLV support:
 - pack and unpack the sub-TLV
 - display data in "show isis database detail"

Signed-off-by: G. Paul Ziemba <paulz@labn.net>
Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: add the prefix-metric support to flex-algo

Add the support for the prefix-metric (M-Flag) in Flex-Algo.

When a Flex-Algo Definition with the M-Flag is elected, prefixes
redistributed from other protocols or IS-IS levels will be advertised
with an additional specific per-algorithm Flex-Algo metric. This
advertisement was implemented in previous commits. This metric is used
instead of the IGP metric when computing paths. If this metric is
unavailable (likely due to an incorrect Flex-Algo implementation on
another system), the IGP metric will not be used, and the prefix will
be dropped.

A prefix is considered redistributed based on the following criteria:

- The prefix is of the IPv6 type and has the external bit set directly
  in the IPv6 Reachability TLV.
- The prefix-SID for the specific algorithm has the Re-advertisement
  flag.
- The prefix has a prefix attribute subTLV with the External flag.

Signed-off-by: G. Paul Ziemba <paulz@labn.net>
Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: add support of prefix attribute flags sub-TLV

Add the support for prefix attribute flags sub-TLV (RFC7794):
 - pack and unpack the sub-TLV
 - display data in "show isis database detail"

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: set prefix metric sub-TLV

Set the prefix metric sub-TLV for external prefix in "Extended IP
reachability" and "IPv6 reachability" TLVs.

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
doc: add flex-algo prefix metric

Add the documentation of the "prefix-metric" Flex-Algo Definition
option.

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind committed:
isisd: set prefix attribute flags

Advertise the prefix attribute flag Sub-TLV for prefixes that are either
of the node, external or redistributed type. No Sub-TLVs with a flag of
0x0 are sent for prefixes that are not of these types.

Note that on some other implementations, the 0x0 flag is always
advertised even if it is not RFC mandatory.

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

odd22 commented on isisd/isis_flex_algo.c:
```diff
@@ -327,4 +329,162 @@ bool isis_flex_algo_constraint_drop(struct isis_spftree *spftree,
+		break;
+	}
+
+	if (asla && CHECK_FLAG(asla->standard_apps, ISIS_SABM_FLAG_X)) {
```
Here, you can't be sure that asla is set or not. First, in declaration asla is not initialized to NULL, second, the previous for loop could lead asla initialized or not depending if the loop finish at the end of the list of with the break.

louis-6wind replied:
ALL_LIST_ELEMENTS_RO initializes asla to NULL and sets it to NULL at the end.
The code is correct but can be clarified

louis-6wind replied:
done

odd22 commented on isisd/isis_lsp.c:
```diff
@@ -2297,21 +2297,39 @@ static int lsp_handle_adj_state_change(struct isis_adjacency *adj)
+ * algorithm parameter: was added to support flex-algo. For non-flex-algo
+ * usage, algorithm should be 0. For flex-algo usage, algorithm should be
+ * in the range 128-255.
  */
```
This is a wrong modification of isis_lsp_iterate_ip_reach(). In fact, these functions was originally defined to ease loop other all LSPs. This is generic functions that could be used in various place and which execute a callback function pass as arguments. It is use in various place (isis_sr.c, isis_te.c at least). For flex algo, instead of modify the generic function, please call the function with appropriate callback function to process Flex algo. The callback function could return LSP_ITER_CONTINUE or LSP_ITER_STOP to control how the iteration on LSP is done.

odd22 reviewed (changes requested):
Good jobs which enhance Flex Algo, but I have some remarks on the code:

- Regarding the documentation and the RFC9350, you should prune link that not support/advertise metric for the corresponding Flex Algo. The method you use, set metric with max value, is not suitable if at the end you need to choose between 2 equal paths that reach the maximum value and second, this lead to degrade performance as you explore path that could not be take into account if they goes through a link that should be prune
- In the documentation, you introduce a new command to manage prefix-metric i.e. M-Flag, but it is not tested. Can you add one more test or tune a router configuration to test it ?
- There is a possible problem when looping asla metric. Please check and fix it
- The way you modify the isis_lsp_iterate_ip_reach() is not good. It is preferable to keep the function simple i.e. loop over LSP and launch a call back function for each entries and put complexity or specific process in the call back function

donaldsharp commented:
can we run `--enable-address-sanitizer` with this change?  the TE code has a bunch of memory leaks and I do not want to add more

louis-6wind commented on doc/user/isisd.rst:
```diff
@@ -530,14 +529,45 @@ The following commands configure Flex-Algo at the 'router isis' and
+   metric is in the narrow format, the maximum wide metric value (2^24 -1) is
+   used. The TE metric refers to the 'metric' value of the 'interface'
+   'link-params' section.
+
```
I agree with you. My mistake

louis-6wind replied:
done

louis-6wind commented on isisd/isis_flex_algo.c:
```diff
@@ -327,4 +329,162 @@ bool isis_flex_algo_constraint_drop(struct isis_spftree *spftree,
+ * IP reachability TLVs for external routes.
+ *
+ * draft-ietf-lsr-flex-alg Section 8
+ * "IS-IS Flexible Algorithm Prefix Metric Sub-TLV"
```
done

louis-6wind committed:
tests: add flex-algo prefix-metric flag test

Test the redistribution of the prefix in a flex-algo topology on which
the prefix-metric flag is set.

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind commented:
> Good jobs which enhance Flex Algo, but I have some remarks on the code:
> 
> * Regarding the documentation and the RFC9350, you should prune link that not support/advertise metric for the corresponding Flex Algo. The method you use, set metric with max value, is not suitable if at the end you need to choose between 2 equal paths that reach the maximum value and second, this lead to degrade performance as you explore path that could not be take into account if they goes through a link that should be prune
was fixed
> * In the documentation, you introduce a new command to manage prefix-metric i.e. M-Flag, but it is not tested. Can you add one more test or tune a router configuration to test it ?
test added
> * There is a possible problem when looping asla metric. Please check and fix it
was fixed
> * The way you modify the isis_lsp_iterate_ip_reach() is not good. It is preferable to keep the function simple i.e. loop over LSP and launch a call back function for each entries and put complexity or specific process in the call back function
I am now calling external functions

louis-6wind committed:
isisd: fix a checkpatch issue

> < WARNING: Block comments use a trailing */ on a separate line

Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

louis-6wind commented:
ci:rerun

louis-6wind commented:
ci:rerun

riw777 commented:
@frrbot autoclose in 2 weeks
